sys/sem/sem.c

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2013-15 Freie Universität Berlin
+ * Copyright (C) 2015 Martine Lenders <mlenders@inf.fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+
+#include "irq.h"
+#include "msg.h"
+#include "vtimer.h"
+
+#include "sem.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define _MSG_SIGNAL         (0x0501)
+#define _MSG_TIMEOUT        (0x0502)
+#define _MSG_DESTROYED      (0x0503)
+
+int sem_create(sem_t *sem, unsigned int value)
+{
+    if (sem == NULL) {
+        return -EINVAL;
+    }
+    sem->value = value;
+    /* waiters for the mutex */
+    sem->queue.first = NULL;
+    return 0;
+}
+
+int sem_destroy(sem_t *sem)
+{
+    unsigned int old_state;
+    priority_queue_node_t *next;
+    if (sem == NULL) {
+        return -EINVAL;
+    }
+    old_state = disableIRQ();
+    next = priority_queue_remove_head(&sem->queue);
+    while (next) {
+        msg_t msg;
+        kernel_pid_t pid = (kernel_pid_t)next->data;
+        msg.type = _MSG_DESTROYED;
+        msg_send_int(&msg, pid);
+    }
+    restoreIRQ(old_state);
+    return 0;
+}
+
+int sem_wait_timed(sem_t *sem, timex_t *timeout)
+{
+    if (sem == NULL) {
+        return -EINVAL;
+    }
+    assert(sched_active_thread->msg_array != NULL);
```

Comment:
This precondition is not needed. Every thread can receive messages, but the sender is immediately message send blocked if the receiver has no (or a full) message queue.